src/metabase/channel/render/util.clj

```diff
@@ -37,6 +37,61 @@
    (and (some? dashcard)
         (get-in dashcard [:visualization_settings :visualization]))))
 
+(defn is-scalar-funnel?
+  "Check if the visualization is a scalar funnel.
+   Matches the frontend implementation in frontend/src/metabase/visualizer/visualizations/funnel.ts"
+  [{:keys [display settings]}]
+  (and (= display "funnel")
+       (= (get settings :funnel.metric) "METRIC")
+       (= (get settings :funnel.dimension) "DIMENSION")))
+
+(defn get-visualization-columns
+  "Creates visualization columns for a visualizer entity.
+   Similar to the frontend implementation in src/metabase/visualizer/utils/get-visualization-columns.ts"
+  [visualizer-definition series-data]
+  (let [{:keys [columnValuesMapping settings] :as viz-def} visualizer-definition]
+    (cond
+      ;; Scalar funnel uses pre-defined metric and dimension columns
+      (is-scalar-funnel? viz-def)
+      (let [metric-column-name (get settings :funnel.metric)
+            dimension-column-name (get settings :funnel.dimension)
+            main-card-with-data (first series-data)
+            base-type (get-in main-card-with-data [:data :cols 0 :base_type])]
+        [{:name metric-column-name
+          :display_name metric-column-name
+          :base_type (or base-type :type/Number)
+          :semantic_type :type/Quantity}
+         {:name dimension-column-name
+          :display_name dimension-column-name
+          :base_type :type/Text
+          :semantic_type :type/Category}])
+
+      ;; For all other chart types, create visualization columns from column mappings
+      ;; TODO: Non scalar visualizer funnels are currently not officially supported
+      :else
+      (reduce
+       (fn [columns [viz-col-name column-mappings]]
+         (concat
+          columns
+          (filter some?
+                  (map
+                   (fn [mapping]
```

Comment:
nit - function gets a bit long, maybe this anonymous function can be broken out with a meaningful name?